objects/src/accounts/component/template/storage/placeholder.rs

```diff
@@ -0,0 +1,184 @@
+/// replace these placeholders with the instance’s actual values.
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct StoragePlaceholder {
+    key: String,
+}
+
+impl StoragePlaceholder {
+    /// Creates a new [StoragePlaceholder] from the provided string.
+    ///
+    /// A [StoragePlaceholder] serves as an identifier for storage values that are determined at
+    /// instantiation time of an [AccountComponentTemplate](super::super::AccountComponentTemplate).
+    ///
+    /// The key can consist of one or more segments separated by dots (`.`).  
+    /// Each segment must be non-empty and may contain only alphanumeric characters, underscores
+    /// (`_`), or hyphens (`-`).
+    ///
+    /// # Errors
+    ///
+    /// This method returns an error if:
+    /// - Any segment (or the whole key) is empty.
+    /// - Any segment contains invalid characters.
+    pub fn new(key: impl Into<String>) -> Result<Self, StoragePlaceholderError> {
+        let key: String = key.into();
+        Self::validate(&key)?;
+        Ok(Self { key })
+    }
+
+    /// Returns the key name
+    pub fn inner(&self) -> &str {
+        &self.key
+    }
+
+    /// Checks if the given string is a valid key.
+    /// A storage placeholder is valid if it's made of one or more segments that are non-empty
+    /// alphanumeric strings.
+    fn validate(key: &str) -> Result<(), StoragePlaceholderError> {
+        if key.is_empty() {
+            return Err(StoragePlaceholderError::EmptyKey);
+        }
+
+        for segment in key.split('.') {
+            if segment.is_empty() {
+                return Err(StoragePlaceholderError::EmptyKey);
+            }
+
+            for c in segment.chars() {
+                if !(c.is_ascii_alphanumeric() || c == '_' || c == '-') {
+                    return Err(StoragePlaceholderError::InvalidChar(key.into(), c));
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl TryFrom<&str> for StoragePlaceholder {
+    type Error = StoragePlaceholderError;
+
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        if value.starts_with("{{") && value.ends_with("}}") {
+            let inner = &value[2..value.len() - 2];
+            Self::validate(inner)?;
+
+            Ok(StoragePlaceholder { key: inner.to_string() })
+        } else {
+            Err(StoragePlaceholderError::FormatError(value.into()))
+        }
+    }
+}
+
+impl TryFrom<&String> for StoragePlaceholder {
+    type Error = StoragePlaceholderError;
+
+    fn try_from(value: &String) -> Result<Self, Self::Error> {
+        Self::try_from(value.as_str())
+    }
+}
+
+impl core::fmt::Display for StoragePlaceholder {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        write!(f, "{{{{{}}}}}", self.key)
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum StoragePlaceholderError {
+    #[error("key segment cannot be empty")]
```

Comment:
```suggestion
    #[error("entire key and key segments cannot be empty")]
```

Nit: This error is used for two error conditions.

Reply:
Seemed like the two conditions could be reduced to just the segment condition (since an empty key is 1 segment long anyway), but in any case your suggestion works much better so I'm changing it